### PR TITLE
Adding ability to upload native coverage files like .coverage, .covx , .covb and .cjson to ADO. 

### PR DESCRIPTION
**Pull request by rasunkar**
**Summary**

1. Adding capabilities to upload native coverage files like .coverage, .covx, .covb, .cjson to tcm log store.
2. This would allow the files to be processed for merging as part of build completion event to generate the final view.

**Local Testing** -

**Setup** 
The edited task is run with a variety of input files, including .coverage, .covx etc. 
Checked that the merged file was being uploaded in logstore

1. **.coverage** 

coverage

2.  **.covx**

covx

3. **.covb**

covb

4. **cjson**

cjson

### CODE REVIEW

File: src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs [hunk 1]
```diff
@@ -165,16 +165,75 @@ public async Task PublishHTMLReport(string reportDirectory, CancellationToken to
...
+        }
+
+        private TestLogType GetTestLogType(string nativeCoverageFile) 
+        { 
+            TestLogType logType = TestLogType.Intermediate;
+        
+            if (nativeCoverageFile != null & 
+                nativeCoverageFile.EndsWith(Constants.CoverageConstants.CoverageFileExtension))
+            {
+                logType = TestLogType.CodeCoverage;
```
Comment (vinayakmsft) [inline src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs:222]: Will this TestLogType.CodeCoverage cover only for .coverage?

File: src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs [hunk 2]
```diff
@@ -165,16 +165,75 @@ public async Task PublishHTMLReport(string reportDirectory, CancellationToken to
...
+                TraceLogger.Error(string.Format(Resources.FailedToUploadNativeCoverageFiles, ex));
+            }
+        }
+
+        private TestLogType GetTestLogType(string nativeCoverageFile) 
+        { 
+            TestLogType logType = TestLogType.Intermediate;
+        
+            if (nativeCoverageFile != null & 
+                nativeCoverageFile.EndsWith(Constants.CoverageConstants.CoverageFileExtension))
```
Comment (vinayakmsft) [inline src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs:220]: Do we need to have conditions for other formats as well like .covx/.covb?

Comment (rasunkar) [reply src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs:220]: only for .coverage we store it as coverage file rest all are intermediate. pls check line 217

File: src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs [hunk 3]
```diff
@@ -165,16 +165,75 @@ public async Task PublishHTMLReport(string reportDirectory, CancellationToken to
...
         }
 
+        public bool IsUploadNativeFilesToTCMSupported()
+        {
+            return _featureFlagHelper.GetFeatureFlagState(Constants.FeatureFlags.UploadNativeCoverageFilesToLogStore, false);
+        }
+
+        public async Task PublishNativeCoverageFiles(IList<string> nativeCoverageFiles, CancellationToken cancellationToken)
+        {
+            if (nativeCoverageFiles.Count == 0)
```
Comment (eswarnathp) [inline src/CoveragePublisher/Publishers/DefaultPublisher/AzurePipelinesPublisher.cs:175]: can we handle null check as well

**Review by vinayakmsft — APPROVED** [review]

File: src/CoveragePublisher/CoverageProcessor.cs [hunk 4]
```diff
@@ -42,11 +42,24 @@ public async Task ParseAndPublishCoverage(PublisherConfiguration config, Cancell
...
-                    if (supportsFileCoverageJson)
+                    if (uploadNativeCoverageFilesToLogStore)
                     {
-                        var fileCoverage = parser.GetFileCoverageInfos();
+                        // Upload native coverage files to TCM
+                        TraceLogger.Debug("Publishing native coverage files is supported.");
+
+                        await _publisher.PublishNativeCoverageFiles(config.CoverageFiles, token);
 
+                        using (new SimpleTimer("CoverageProcesser", "PublishFileCoverage", _telemetry))
```
Comment (vinayakmsft) [inline src/CoveragePublisher/CoverageProcessor.cs:56]: As the fileCoverage object will be null for this IF condition, we dont need this PublishFileCoverage method. Only PublishNativeCoverageFiles will should be fine

File: src/CoveragePublisher/CoverageProcessor.cs [hunk 5]
```diff
@@ -73,13 +86,6 @@ public async Task ParseAndPublishCoverage(PublisherConfiguration config, Cancell
                         {
                             TraceLogger.Warning(Resources.NoCoverageFilesGenerated);
                         }
-                        else
```
Comment (vinayakmsft) [inline src/CoveragePublisher/CoverageProcessor.cs:76]: We should have PublishFileCoverage here, no changes to be made in the IF (supportFileCoverageJson) block

**Review by vinayakmsft — APPROVED** [review]

**Pull request merged**